feat_: Add system time change detection and handling

qfrank opened the pull request:
- Implement time change detection mechanism in NTPTimeSource
- Add TimeChangeThreshold constant to define significant time shifts
- Update Now() method to track and detect system time changes
- Log warnings and attempt to update NTP offset when time inconsistency is detected
- Initialize time tracking fields during Start() method

this PR should fix [mobile issue](https://github.com/status-im/status-mobile/issues/17739)

relate mobile [PR](https://github.com/status-im/status-mobile/pull/22178)

qfrank committed:
fix_: improve clock value validation and time synchronization

- Refactor clock value generation in chat protocol
- Enhance time source offset handling
- Expose ValidateClockValue method for broader use
- Add logging for clock value validation errors
- Improve time synchronization retry mechanism

qfrank commented on protocol/chat.go:
```diff
@@ -400,15 +404,34 @@ func (c *Chat) updateChatFromGroupMembershipChanges(g *v1protocol.Group) {
+		}
+
+		// TODO: validate clock value again and return error if still invalid
+	}
```
Will create the issue and add link here once this PR get approved

ilmotta replied:
Could you elaborate on why you prefer to add the secondary check in a separate PR? I'm asking this because I see the secondary validation as part of a coherent solution of this PR. This situation can arise for instance if the network is unstable and the reply from NTP servers timeout before `DefaultRPCTimeout`.

qfrank replied:
I thought many places used `NextClockAndTimestamp` and I wanted to make a quick fix.

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (142)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 0e020993 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/1/) | 2025-02-24 06:13:26 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250224-061034-0e0209-pr6358.zip) |
| :heavy_check_mark: | 0e020993 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/1/) | 2025-02-24 06:13:34 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250224-061033-0e0209-pr6358.aar) |
| :heavy_check_mark: | 0e020993 | [#1](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/1/) | 2025-02-24 06:14:21 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250224-061039-0e0209-pr6358.zip) |
| :heavy_check_mark: | 0e020993 | [#1](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/1/) | 2025-02-24 06:15:27 | ~4 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250224-061040-0e0209-pr6358.zip) |
| :heavy_check_mark: | 0e020993 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/1/) | 2025-02-24 06:16:00 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250224-061035-0e0209-pr6358.zip) |
| :heavy_check_mark: | 0e020993 | [#1](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/1/) | 2025-02-24 06:16:17 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250224-061039-0e0209-pr6358.zip) |
| :heavy_check_mark: | 0e020993 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/1/) | 2025-02-24 06:24:36 | ~13 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/1//consoleText) |
| :heavy_multiplication_x: | 0e020993 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/1/) | 2025-02-24 06:42:55 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 25699169 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/2/) | 2025-02-24 07:40:19 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250224-073724-256991-pr6358.zip) |
| :heavy_check_mark: | 25699169 | [#2](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/2/) | 2025-02-24 07:40:30 | ~2 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250224-073732-256991-pr6358.zip) |
| :heavy_check_mark: | 25699169 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/2/) | 2025-02-24 07:41:56 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250224-073724-256991-pr6358.aar) |
| :heavy_check_mark: | 25699169 | [#2](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/2/) | 2025-02-24 07:42:27 | ~4 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250224-073727-256991-pr6358.zip) |
| :heavy_check_mark: | 25699169 | [#2](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/2/) | 2025-02-24 07:42:54 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250224-073728-256991-pr6358.zip) |
| :heavy_check_mark: | 25699169 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/2/) | 2025-02-24 07:43:02 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250224-073726-256991-pr6358.zip) |
| :heavy_check_mark: | 25699169 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/2/) | 2025-02-24 07:50:05 | ~12 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/2//consoleText) |
| :heavy_check_mark: | 25699169 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/2/) | 2025-02-24 08:09:47 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 6a156c66 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/3/) | 2025-03-03 06:55:15 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250303-065227-6a156c-pr6358.zip) |
| :heavy_check_mark: | 6a156c66 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/3/) | 2025-03-03 06:55:24 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250303-065225-6a156c-pr6358.aar) |
| :heavy_check_mark: | 6a156c66 | [#3](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/3/) | 2025-03-03 06:56:31 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250303-065231-6a156c-pr6358.zip) |
| :heavy_check_mark: | 6a156c66 | [#3](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/3/) | 2025-03-03 06:57:27 | ~4 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250303-065235-6a156c-pr6358.zip) |
| :heavy_check_mark: | 6a156c66 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/3/) | 2025-03-03 06:57:50 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250303-065227-6a156c-pr6358.zip) |
| :heavy_check_mark: | 6a156c66 | [#3](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/3/) | 2025-03-03 06:58:09 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250303-065231-6a156c-pr6358.zip) |
| :heavy_check_mark: | 6a156c66 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/3/) | 2025-03-03 07:06:00 | ~13 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/3//consoleText) |
| :heavy_check_mark: | 6a156c66 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/3/) | 2025-03-03 07:25:03 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 660ad7a1 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/4/) | 2025-03-05 05:46:45 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-054404-660ad7-pr6358.zip) |
| :heavy_check_mark: | 660ad7a1 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/4/) | 2025-03-05 05:46:58 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-054402-660ad7-pr6358.aar) |
| :heavy_multiplication_x: | 660ad7a1 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/4/) | 2025-03-05 05:47:14 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/4//consoleText) |
| :heavy_check_mark: | 660ad7a1 | [#4](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/4/) | 2025-03-05 05:47:29 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-054412-660ad7-pr6358.zip) |
| :heavy_check_mark: | 660ad7a1 | [#4](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/4/) | 2025-03-05 05:48:08 | ~4 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-054408-660ad7-pr6358.zip) |
| :heavy_check_mark: | 660ad7a1 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/4/) | 2025-03-05 05:49:36 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-054405-660ad7-pr6358.zip) |
| :heavy_check_mark: | 660ad7a1 | [#4](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/4/) | 2025-03-05 05:49:38 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-054408-660ad7-pr6358.zip) |
| :heavy_check_mark: | 660ad7a1 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/4/) | 2025-03-05 05:56:20 | ~12 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | f14eb618 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/5/) | 2025-03-05 07:37:45 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-073459-f14eb6-pr6358.zip) |
| :heavy_check_mark: | f14eb618 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/5/) | 2025-03-05 07:37:55 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-073458-f14eb6-pr6358.aar) |
| :heavy_multiplication_x: | f14eb618 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/5/) | 2025-03-05 07:38:08 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/5//consoleText) |
| :heavy_check_mark: | f14eb618 | [#5](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/5/) | 2025-03-05 07:38:27 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-073507-f14eb6-pr6358.zip) |
| :heavy_check_mark: | f14eb618 | [#5](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/5/) | 2025-03-05 07:38:57 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-073503-f14eb6-pr6358.zip) |
| :heavy_check_mark: | f14eb618 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/5/) | 2025-03-05 07:40:26 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-073501-f14eb6-pr6358.zip) |
| :heavy_check_mark: | f14eb618 | [#5](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/5/) | 2025-03-05 07:40:53 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-073503-f14eb6-pr6358.zip) |
| :heavy_check_mark: | f14eb618 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/5/) | 2025-03-05 07:47:12 | ~12 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | baac4ed3 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/6/) | 2025-03-05 07:45:08 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-074224-baac4e-pr6358.zip) |
| :heavy_check_mark: | baac4ed3 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/6/) | 2025-03-05 07:45:19 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-074223-baac4e-pr6358.aar) |
| :heavy_multiplication_x: | baac4ed3 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/6/) | 2025-03-05 07:45:31 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/6//consoleText) |
| :heavy_check_mark: | baac4ed3 | [#6](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/6/) | 2025-03-05 07:45:50 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-074231-baac4e-pr6358.zip) |
| :heavy_check_mark: | baac4ed3 | [#6](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/6/) | 2025-03-05 07:46:19 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-074227-baac4e-pr6358.zip) |
| :heavy_check_mark: | baac4ed3 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/6/) | 2025-03-05 07:47:44 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-074225-baac4e-pr6358.zip) |
| :heavy_check_mark: | baac4ed3 | [#6](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/6/) | 2025-03-05 07:48:18 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-074227-baac4e-pr6358.zip) |
| | | | | | | |
| :heavy_check_mark: | 3d588b30 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/7/) | 2025-03-05 07:47:54 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-074512-3d588b-pr6358.zip) |
| :heavy_multiplication_x: | 3d588b30 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/7/) | 2025-03-05 07:48:07 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/7//consoleText) |
| :heavy_check_mark: | 3d588b30 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/7/) | 2025-03-05 07:48:17 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-074523-3d588b-pr6358.aar) |
| :heavy_check_mark: | 3d588b30 | [#7](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/7/) | 2025-03-05 07:49:42 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-074557-3d588b-pr6358.zip) |
| :heavy_check_mark: | 3d588b30 | [#7](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/7/) | 2025-03-05 07:50:23 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-074624-3d588b-pr6358.zip) |
| :heavy_check_mark: | 3d588b30 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/7/) | 2025-03-05 07:53:07 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-074748-3d588b-pr6358.zip) |
| :heavy_check_mark: | 3d588b30 | [#7](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/7/) | 2025-03-05 07:53:58 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-074826-3d588b-pr6358.zip) |
| :heavy_check_mark: | 3d588b30 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/6/) | 2025-03-05 07:59:04 | ~11 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/6//consoleText) |
| | | | | | | |
| :heavy_check_mark: | bc36f945 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/8/) | 2025-03-05 08:03:12 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-080031-bc36f9-pr6358.zip) |
| :heavy_check_mark: | bc36f945 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/8/) | 2025-03-05 08:03:23 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-080029-bc36f9-pr6358.aar) |
| :heavy_check_mark: | bc36f945 | [#8](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/8/) | 2025-03-05 08:03:39 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-080038-bc36f9-pr6358.zip) |
| :heavy_check_mark: | bc36f945 | [#8](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/8/) | 2025-03-05 08:04:27 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-080034-bc36f9-pr6358.zip) |
| :heavy_check_mark: | bc36f945 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/8/) | 2025-03-05 08:05:50 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-080033-bc36f9-pr6358.zip) |
| :heavy_check_mark: | bc36f945 | [#8](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/8/) | 2025-03-05 08:05:59 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-080034-bc36f9-pr6358.zip) |
| :heavy_check_mark: | bc36f945 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/7/) | 2025-03-05 08:13:34 | ~12 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/7//consoleText) |
| :heavy_check_mark: | bc36f945 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/8/) | 2025-03-05 08:31:39 | ~31 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/8//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 8445de58 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/9/) | 2025-03-05 08:43:39 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-084038-8445de-pr6358.aar) |
| :heavy_check_mark: | 8445de58 | [#9](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/9/) | 2025-03-05 08:43:50 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-084047-8445de-pr6358.zip) |
| :heavy_check_mark: | 8445de58 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/9/) | 2025-03-05 08:44:10 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-084039-8445de-pr6358.zip) |
| :heavy_check_mark: | 8445de58 | [#9](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/9/) | 2025-03-05 08:44:40 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-084041-8445de-pr6358.zip) |
| :heavy_check_mark: | 8445de58 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/9/) | 2025-03-05 08:46:00 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-084039-8445de-pr6358.zip) |
| :heavy_check_mark: | 8445de58 | [#9](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/9/) | 2025-03-05 08:46:11 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-084043-8445de-pr6358.zip) |
| :heavy_check_mark: | 8445de58 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/8/) | 2025-03-05 08:53:30 | ~12 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/8//consoleText) |
| :heavy_check_mark: | 8445de58 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/9/) | 2025-03-05 09:11:50 | ~31 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/9//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | bcd5afa0 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/10/) | 2025-03-05 09:14:52 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/10//consoleText) |
| :heavy_check_mark: | bcd5afa0 | [#10](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/10/) | 2025-03-05 09:15:02 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-091143-bcd5af-pr6358.zip) |
| :heavy_check_mark: | bcd5afa0 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/10/) | 2025-03-05 09:15:17 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-091134-bcd5af-pr6358.aar) |
| :heavy_check_mark: | bcd5afa0 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/10/) | 2025-03-05 09:17:27 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-091136-bcd5af-pr6358.zip) |
| :heavy_check_mark: | bcd5afa0 | [#10](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/10/) | 2025-03-05 09:17:35 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-091139-bcd5af-pr6358.zip) |
| :heavy_check_mark: | bcd5afa0 | [#10](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/10/) | 2025-03-05 09:20:05 | ~8 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-091138-bcd5af-pr6358.zip) |
| :heavy_check_mark: | bcd5afa0 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/9/) | 2025-03-05 09:23:52 | ~12 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/9//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 2bc9c8a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/11/) | 2025-03-05 09:21:24 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-091816-2bc9c8-pr6358.aar) |
| :heavy_check_mark: | 2bc9c8a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/11/) | 2025-03-05 09:21:51 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-091825-2bc9c8-pr6358.zip) |
| :heavy_multiplication_x: | 2bc9c8a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/11/) | 2025-03-05 09:22:43 | ~4 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/11//consoleText) |
| :heavy_check_mark: | 2bc9c8a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/11/) | 2025-03-05 09:23:44 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-091822-2bc9c8-pr6358.zip) |
| :heavy_check_mark: | 2bc9c8a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/11/) | 2025-03-05 09:24:01 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-091820-2bc9c8-pr6358.zip) |
| :heavy_check_mark: | 2bc9c8a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/11/) | 2025-03-05 09:26:48 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-092253-2bc9c8-pr6358.zip) |
| :heavy_check_mark: | 2bc9c8a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/11/) | 2025-03-05 09:27:27 | ~7 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-092018-2bc9c8-pr6358.zip) |
| :heavy_check_mark: | 2bc9c8a3 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/10/) | 2025-03-05 09:36:31 | ~12 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/10//consoleText) |
| | | | | | | |
| :heavy_check_mark: | c5f587f1 | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/12/) | 2025-03-05 09:28:05 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-092455-c5f587-pr6358.aar) |
| :heavy_check_mark: | c5f587f1 | [#12](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/12/) | 2025-03-05 09:28:12 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-092503-c5f587-pr6358.zip) |
| :heavy_check_mark: | c5f587f1 | [#12](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/12/) | 2025-03-05 09:30:20 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-092459-c5f587-pr6358.zip) |
| :heavy_check_mark: | c5f587f1 | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/12/) | 2025-03-05 09:30:27 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-092457-c5f587-pr6358.zip) |
| :heavy_multiplication_x: | c5f587f1 | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/12/) | 2025-03-05 09:31:28 | ~6 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/12//consoleText) |
| :heavy_check_mark: | c5f587f1 | [#12](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/12/) | 2025-03-05 09:34:44 | ~7 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-092733-c5f587-pr6358.zip) |
| :heavy_check_mark: | c5f587f1 | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/12/) | 2025-03-05 09:45:53 | ~18 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-092652-c5f587-pr6358.zip) |
| :heavy_check_mark: | c5f587f1 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/11/) | 2025-03-05 09:50:07 | ~13 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/11//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 1b07628c | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/13/) | 2025-03-05 23:46:28 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250305-234349-1b0762-pr6358.zip) |
| :heavy_multiplication_x: | 1b07628c | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/13/) | 2025-03-05 23:46:54 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/13//consoleText) |
| :heavy_check_mark: | 1b07628c | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/13/) | 2025-03-05 23:47:01 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250305-234348-1b0762-pr6358.aar) |
| :heavy_check_mark: | 1b07628c | [#13](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/13/) | 2025-03-05 23:47:40 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250305-234357-1b0762-pr6358.zip) |
| :heavy_check_mark: | 1b07628c | [#13](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/13/) | 2025-03-05 23:47:50 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-234353-1b0762-pr6358.zip) |
| :heavy_check_mark: | 1b07628c | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/13/) | 2025-03-05 23:49:11 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250305-234351-1b0762-pr6358.zip) |
| :heavy_check_mark: | 1b07628c | [#13](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/13/) | 2025-03-05 23:49:19 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250305-234353-1b0762-pr6358.zip) |
| :heavy_check_mark: | 1b07628c | [#12](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/12/) | 2025-03-05 23:57:46 | ~13 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/12//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 6d80410e | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/14/) | 2025-03-11 11:06:59 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250311-110421-6d8041-pr6358.zip) |
| :heavy_check_mark: | 6d80410e | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/14/) | 2025-03-11 11:07:14 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250311-110420-6d8041-pr6358.aar) |
| :heavy_check_mark: | 6d80410e | [#14](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/14/) | 2025-03-11 11:08:02 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250311-110423-6d8041-pr6358.zip) |
| :heavy_check_mark: | 6d80410e | [#14](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/14/) | 2025-03-11 11:08:48 | ~4 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250311-110428-6d8041-pr6358.zip) |
| :heavy_check_mark: | 6d80410e | [#14](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/14/) | 2025-03-11 11:09:47 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250311-110425-6d8041-pr6358.zip) |
| :heavy_check_mark: | 6d80410e | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/14/) | 2025-03-11 11:09:57 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250311-110423-6d8041-pr6358.zip) |
| :heavy_check_mark: | 6d80410e | [#13](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/13/) | 2025-03-11 11:21:55 | ~17 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/13//consoleText) |
| :heavy_multiplication_x: | 6d80410e | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/14/) | 2025-03-11 11:36:38 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/14//consoleText) |
| | | | | | | |
| :heavy_check_mark: | bb7b7c32 | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/15/) | 2025-03-12 02:55:06 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250312-025223-bb7b7c-pr6358.zip) |
| :heavy_check_mark: | bb7b7c32 | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/15/) | 2025-03-12 02:55:18 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250312-025222-bb7b7c-pr6358.aar) |
| :heavy_check_mark: | bb7b7c32 | [#15](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/15/) | 2025-03-12 02:55:59 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250312-025230-bb7b7c-pr6358.zip) |
| :heavy_check_mark: | bb7b7c32 | [#15](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/15/) | 2025-03-12 02:56:09 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250312-025227-bb7b7c-pr6358.zip) |
| :heavy_check_mark: | bb7b7c32 | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/15/) | 2025-03-12 02:57:45 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250312-025224-bb7b7c-pr6358.zip) |
| :heavy_check_mark: | bb7b7c32 | [#15](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/15/) | 2025-03-12 02:57:55 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250312-025227-bb7b7c-pr6358.zip) |
| :heavy_check_mark: | bb7b7c32 | [#14](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/14/) | 2025-03-12 03:16:41 | ~24 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/14//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 8aa6dab7 | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/16/) | 2025-03-12 03:17:39 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250312-031459-8aa6da-pr6358.zip) |
| :heavy_multiplication_x: | 8aa6dab7 | [#16](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/16/) | 2025-03-12 03:17:41 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/16//consoleText) |
| :heavy_check_mark: | 8aa6dab7 | [#16](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/16/) | 2025-03-12 03:17:54 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250312-031459-8aa6da-pr6358.aar) |
| :heavy_check_mark: | 8aa6dab7 | [#16](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/16/) | 2025-03-12 03:18:24 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250312-031507-8aa6da-pr6358.zip) |
| :heavy_check_mark: | 8aa6dab7 | [#16](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/16/) | 2025-03-12 03:18:37 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250312-031503-8aa6da-pr6358.zip) |
| :heavy_check_mark: | 8aa6dab7 | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/16/) | 2025-03-12 03:20:24 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250312-031501-8aa6da-pr6358.zip) |
| :heavy_check_mark: | 8aa6dab7 | [#16](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/16/) | 2025-03-12 03:20:53 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250312-031503-8aa6da-pr6358.zip) |
| :heavy_check_mark: | 8aa6dab7 | [#15](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/15/) | 2025-03-12 03:36:14 | ~19 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/15//consoleText) |
| :heavy_check_mark: | 8aa6dab7 | [#17](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/17/) | 2025-03-12 04:04:24 | ~32 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/17//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | b062a297 | [#18](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/18/) | 2025-03-13 11:24:35 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/18//consoleText) |
| :heavy_check_mark: | b062a297 | [#17](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/17/) | 2025-03-13 11:25:20 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250313-112234-b062a2-pr6358.zip) |
| :heavy_check_mark: | b062a297 | [#17](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/17/) | 2025-03-13 11:25:32 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250313-112233-b062a2-pr6358.aar) |
| :heavy_check_mark: | b062a297 | [#17](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/17/) | 2025-03-13 11:25:49 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250313-112242-b062a2-pr6358.zip) |
| :heavy_check_mark: | b062a297 | [#17](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/17/) | 2025-03-13 11:26:34 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250313-112237-b062a2-pr6358.zip) |
| :heavy_multiplication_x: | b062a297 | [#19](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/19/) | 2025-03-13 11:27:00 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/19//consoleText) |
| :heavy_check_mark: | b062a297 | [#17](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/17/) | 2025-03-13 11:28:00 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250313-112235-b062a2-pr6358.zip) |
| :heavy_check_mark: | b062a297 | [#17](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/17/) | 2025-03-13 11:28:19 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250313-112238-b062a2-pr6358.zip) |
| :heavy_check_mark: | b062a297 | [#16](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/16/) | 2025-03-13 11:39:58 | ~17 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/16//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 52083ba5 | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/18/) | 2025-03-13 11:32:31 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250313-112949-52083b-pr6358.zip) |
| :heavy_check_mark: | 52083ba5 | [#18](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/18/) | 2025-03-13 11:32:45 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250313-112949-52083b-pr6358.aar) |
| :heavy_check_mark: | 52083ba5 | [#18](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/18/) | 2025-03-13 11:33:04 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250313-112957-52083b-pr6358.zip) |
| :heavy_check_mark: | 52083ba5 | [#18](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/18/) | 2025-03-13 11:33:51 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250313-112952-52083b-pr6358.zip) |
| :heavy_multiplication_x: | 52083ba5 | [#20](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/20/) | 2025-03-13 11:33:59 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/20//consoleText) |
| :heavy_check_mark: | 52083ba5 | [#18](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/18/) | 2025-03-13 11:35:10 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250313-112950-52083b-pr6358.zip) |
| :heavy_check_mark: | 52083ba5 | [#18](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/18/) | 2025-03-13 11:35:24 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250313-112953-52083b-pr6358.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | a6375748 | [#19](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/19/) | 2025-03-13 11:42:31 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250313-113947-a63757-pr6358.zip) |
| :heavy_check_mark: | a6375748 | [#19](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/19/) | 2025-03-13 11:42:40 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250313-113945-a63757-pr6358.aar) |
| :heavy_check_mark: | a6375748 | [#19](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/19/) | 2025-03-13 11:42:51 | ~2 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250313-113953-a63757-pr6358.zip) |
| :heavy_check_mark: | a6375748 | [#19](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/19/) | 2025-03-13 11:43:20 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250313-113950-a63757-pr6358.zip) |
| :heavy_check_mark: | a6375748 | [#19](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/19/) | 2025-03-13 11:45:09 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250313-113948-a63757-pr6358.zip) |
| :heavy_check_mark: | a6375748 | [#19](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/19/) | 2025-03-13 11:45:19 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250313-113951-a63757-pr6358.zip) |
| :heavy_check_mark: | a6375748 | [#17](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/17/) | 2025-03-13 11:59:39 | ~19 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/17//consoleText) |
| :heavy_check_mark: | a6375748 | [#21](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/21/) | 2025-03-13 12:14:03 | ~34 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/21//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 03e1cfb3 | [#20](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-6358/20/) | 2025-03-14 02:53:01 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-250314-025023-03e1cf-pr6358.zip) |
| :heavy_check_mark: | 03e1cfb3 | [#20](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-6358/20/) | 2025-03-14 02:53:13 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-250314-025021-03e1cf-pr6358.aar) |
| :heavy_check_mark: | 03e1cfb3 | [#20](https://ci.status.im/job/status-go/job/prs/job/windows/job/x86_64/job/main/job/PR-6358/20/) | 2025-03-14 02:53:44 | ~3 min | `windows` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-windows-250314-025030-03e1cf-pr6358.zip) |
| :heavy_check_mark: | 03e1cfb3 | [#20](https://ci.status.im/job/status-go/job/prs/job/macos/job/aarch64/job/main/job/PR-6358/20/) | 2025-03-14 02:53:59 | ~3 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250314-025025-03e1cf-pr6358.zip) |
| :heavy_check_mark: | 03e1cfb3 | [#20](https://ci.status.im/job/status-go/job/prs/job/linux/job/x86_64/job/main/job/PR-6358/20/) | 2025-03-14 02:55:53 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-linux-250314-025024-03e1cf-pr6358.zip) |
| :heavy_check_mark: | 03e1cfb3 | [#20](https://ci.status.im/job/status-go/job/prs/job/macos/job/x86_64/job/main/job/PR-6358/20/) | 2025-03-14 02:55:54 | ~5 min | `macos` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-macos-250314-025025-03e1cf-pr6358.zip) |
| :heavy_check_mark: | 03e1cfb3 | [#18](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/18/) | 2025-03-14 03:20:37 | ~30 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-6358/18//consoleText) |
| :heavy_multiplication_x: | 03e1cfb3 | [#23](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/23/) | 2025-03-14 04:21:02 | ~33 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/23//consoleText) |
| :heavy_check_mark: | 03e1cfb3 | [#24](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/24/) | 2025-03-14 05:55:43 | ~35 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/24//consoleText) |
| :heavy_check_mark: | 03e1cfb3 | [#25](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/25/) | 2025-03-14 08:35:15 | ~33 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-6358/25//consoleText) |

qfrank committed:
test_: fix flaky test

ilmotta reviewed (approved):
Thanks for this PR @qfrank.

In terms of the user problem we're trying to mitigate, it occurred to me that if the code detects a clock drift greater than `maxWhisperFutureDriftMs`, the best next step might be to notify the user.

My thinking is, we return an error from a status-go function, cool, but then what should happen to the user experience?

The problem I faced, and the one @churik faced was similar: we couldn't tell from the app UI that there was a clock issue because the difference was too small for us to even consider it as a possible cause.

This may seem like overkill, but could be as simple as a signal emitted from status-go throttled per 30min/1h/etc, which the client can then handle to display a toast or something simple (I have other ideas).

qfrank committed:
feat(timesource)_: Add system time change detection and handling

- Implement time change detection mechanism in NTPTimeSource
- Add TimeChangeThreshold constant to define significant time shifts
- Update Now() method to track and detect system time changes
- Log warnings and attempt to update NTP offset when time inconsistency is detected
- Initialize time tracking fields during Start() method

qfrank committed:
feat(timesource)_: Add TBD comment

qfrank committed:
revert_: "fix_: improve clock value validation and time synchronization"

This reverts commit 3171b12fe43f5aadbc8bc18e6c32e384c4cef2e1.

qfrank committed:
chore_: protect LastClockValue

qfrank commented:
Hi @ilmotta , sorry for the delay as the issue is a bit tricky to handle and I got new ideas, pls review again! previous validation doesn't make much sense as it's not the root cause.

qfrank committed:
fix_: lint issue

ilmotta commented on timesource/timesource_test.go:
```diff
@@ -305,3 +306,280 @@ func TestGetCurrentTimeOffline(t *testing.T) {
+func TestSystemTimeChangeDetection(t *testing.T) {
+	// Create a controlled time source with fixed time
+	currentTime := time.Now()
+	timeJump := 2 * time.Second // Greater than TimeChangeThreshold (1s)
```
`TimeChangeThreshold` could be used directly here instead of the comment.

qfrank replied:
fixed

ilmotta commented on timesource/timesource.go:
```diff
@@ -144,19 +148,57 @@ type NTPTimeSource struct {
+
+		// Synchronously update the NTP offset
+		if err := s.updateOffset(); err != nil {
+			logutils.ZapLogger().Error("Failed to update time offset after time change", zap.Error(err))
```
This log call seems to be redundant because `updateOffset` already logs on error.

qfrank replied:
fixed

ilmotta commented on timesource/timesource.go:
```diff
@@ -28,6 +28,10 @@ const (
+
+	// TimeChangeThreshold defines the minimum time difference that indicates
+	// system time has been changed. Values smaller than this are considered normal drift.
+	TimeChangeThreshold = 1 * time.Second
```
Could you elaborate and comment on source why 1 second seems like a good choice? It's normal that computers drift a few seconds per week if not synced with time servers. https://security.stackexchange.com/questions/227002/is-it-safe-to-assume-that-my-computers-clock-will-always-be-synced-with-actual

qfrank replied:
> TimeChangeThreshold is mainly for detecting whether there has been a manual modification of the system time. Would a computer's clock drift rate affect the calculation of timeDiff.Abs()?

I asked AI above question and following is the answer. 🙂

# TimeChangeThreshold and Clock Drift Analysis

## How the Time Change Detection Works

Looking at your code snippet, the mechanism used to detect system time changes is quite clever:

```go:timesource/timesource.go
// ... existing code ...
monotonicElapsed := time.Since(s.lastMonotonic)
wallClockElapsed := time.Duration(currentTime.UnixNano() - s.lastWallTime.UnixNano())
timeDiff := monotonicElapsed - wallClockElapsed
// ... existing code ...
```

This code uses two different time measurements:

1. **monotonicElapsed**: Measures elapsed time using Go's monotonic clock, which continues steadily and is unaffected by system time changes
2. **wallClockElapsed**: Measures elapsed time using the system wall clock, which can be affected by manual adjustments

## Impact of Clock Drift on the Detection

**The computer's clock drift rate does not significantly impact the `timeDiff.Abs()` calculation** because:

1. Both the monotonic clock and wall clock run on the same hardware, so they drift at essentially the same rate
2. Drift is a gradual process affecting both clocks simultaneously
3. This mechanism doesn't detect absolute time accuracy but rather the relative difference between these two clock sources

## When The Threshold Is Triggered

The threshold will only be exceeded when:
- Someone manually adjusts the system time
- NTP makes a large correction
- The system resumes from sleep/hibernation
- Other events that cause wall clock to suddenly change

## Conclusion

1. The `TimeChangeThreshold` is indeed primarily for detecting manual or sudden system time changes, not for detecting normal clock drift
2. Normal clock drift won't cause significant differences between `monotonicElapsed` and `wallClockElapsed`
3. The 1-second threshold is reasonable for this detection mechanism since it's only sensitive to sudden time changes, not gradual drift

This is a well-designed approach that effectively distinguishes between normal clock drift and artificial time adjustments.

ilmotta replied:
Thank you and thanks bot ;)

qfrank committed:
chore_: address review feedback

qfrank committed:
chore_: fix test

ilmotta commented on timesource/timesource.go:
```diff
@@ -144,19 +148,55 @@ type NTPTimeSource struct {
+	s.timeDataMu.RUnlock()
+
+	// If significant time change detected, update offset synchronously
+	if timeDiff.Abs() > TimeChangeThreshold {
```
Since `timeDiff` depends on usage of the mutex, would it be better to `RUnlock` only after reading `timeDiff`?

qfrank replied:
no, timeDiff already get its value before `s.timeDataMu.RUnlock()`, so it's safe to read after `s.timeDataMu.RUnlock()` , the lock is mainly ensure we get the right value from lastMonotonic / lastWallTime to get correct timeDiff

ilmotta commented on timesource/timesource.go:
```diff
@@ -166,10 +206,12 @@ func (s *NTPTimeSource) updateOffset() error {
 	logutils.ZapLogger().Info("Difference with ntp servers", zap.Duration("offset", offset))
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.timeDataMu.Lock()
```
@qfrank, I'm no expert in managing locks, but in trying to avoid nesting lock management, I thought about modifying `updateOffset` to be just `computeOffset` and logging (almost pure) and then leave the actual field update to the parent code that already manages the mutex. Nothing more than an idea I wanted to explore, but in general I think there are ways to design the code so was to minimize the usage mutexes by making more functions pure (or semi-pure with logging).

Whether you like it or not up to you, but wanted to share in any case.

```diff
modified   timesource/timesource.go
@@ -187,12 +187,11 @@ func (s *NTPTimeSource) Now() time.Time {
 			zap.Duration("threshold", TimeChangeThreshold))
 
 		// Synchronously update the NTP offset
-		if err := s.updateOffset(); err != nil {
-			logutils.ZapLogger().Error("Failed to update time offset after time change", zap.Error(err))
-		}
+		newOffset, _ := s.computeOffset()
 
 		// Update the reference times only after significant time change
 		s.timeDataMu.Lock()
+		s.latestOffset = newOffset
 		s.lastMonotonic = s.now()
 		s.lastWallTime = s.now()
 		s.timeDataMu.Unlock()
@@ -201,20 +200,19 @@ func (s *NTPTimeSource) Now() time.Time {
 	return adjustedTime
 }
 
-func (s *NTPTimeSource) updateOffset() error {
+func (s *NTPTimeSource) computeOffset() (time.Duration, error) {
 	offset, err := computeOffset(s.timeQuery, s.servers, s.allowedFailures)
 	if err != nil {
 		logutils.ZapLogger().Error("failed to compute offset", zap.Error(err))
-		return errUpdateOffset
+		return 0, errUpdateOffset
 	}
 	logutils.ZapLogger().Info("Difference with ntp servers", zap.Duration("offset", offset))
-	s.timeDataMu.Lock()
-	defer s.timeDataMu.Unlock()
-	s.latestOffset = offset
-	//TBD: if we found offset is too large, we should notify user that system time might not be accurate via emit signal,
-	// and because go-waku doesn't use NTPTimeSource ATM (it just use time.Now()), this might be a problem for MissingMessageVerifier work normally.
-	// e.g. might get errInvalidTimeRange when validate StoreQueryRequest
-	return nil
+	// TBD: if we found offset is too large, we should notify user that system time
+	// might not be accurate via emit signal, and because go-waku doesn't use
+	// NTPTimeSource ATM (it just use time.Now()), this might be a problem for
+	// MissingMessageVerifier work normally. e.g. might get errInvalidTimeRange
+	// when validate StoreQueryRequest
+	return offset, nil
 }
 
 // runPeriodically runs periodically the given function based on NTPTimeSource
@@ -260,15 +258,15 @@ func (s *NTPTimeSource) Start() {
 	s.lastMonotonic = currentTime
 	s.lastWallTime = currentTime
 
-	// Attempt to update the offset synchronously so that user can have reliable messages right away
-	err := s.updateOffset()
-	if err != nil {
-		// Failure to update can occur if the node is offline.
-		// Instead of returning an error, continue with the process as the update will be retried periodically.
-		logutils.ZapLogger().Error("failed to update offset", zap.Error(err))
-	}
+	newOffset, err := s.computeOffset()
+	s.latestOffset = newOffset
 
-	s.runPeriodically(s.updateOffset, err == nil)
+	s.runPeriodically(func() error {
+		offset, err := s.computeOffset()
+		s.timeDataMu.Lock()
+		s.latestOffset = offset
+		s.timeDataMu.Unlock()
+		return err
+	}, err == nil)
 
 	s.started = true
 }
modified   timesource/timesource_test.go
@@ -180,7 +180,8 @@ func TestNTPTimeSource(t *testing.T) {
 				now:             time.Now,
 			}
 			assert.WithinDuration(t, time.Now(), source.Now(), clockCompareDelta)
-			err := source.updateOffset()
+			offset, err := source.computeOffset()
+			source.latestOffset = offset
 			if tc.expectError {
 				assert.Equal(t, errUpdateOffset, err)
 			} else {
```

osmaczko commented on timesource/timesource.go:
```diff
@@ -144,19 +148,55 @@ type NTPTimeSource struct {
+
+		// Update the reference times only after significant time change
+		s.timeDataMu.Lock()
+		s.lastMonotonic = s.now()
```
Probably I missed something but it seems `s.lastMonotonic` and `s.lastWallTime` are always equal, as they are updated with the same values in the same places :thinking:
In `func (s *NTPTimeSource) Now() time.Time`
```go
	s.lastMonotonic = s.now()
	s.lastWallTime = s.now()
```
and in `func (s *NTPTimeSource) Start()`
```go
	s.lastMonotonic = currentTime
	s.lastWallTime = currentTime
```

qfrank replied:
yes, they are always the same, that's expected, the most important caculation is here:
```go
	monotonicElapsed := time.Since(s.lastMonotonic)
	wallClockElapsed := time.Duration(currentTime.UnixNano() - s.lastWallTime.UnixNano())
	timeDiff := monotonicElapsed - wallClockElapsed
```	
one field should be enough, I changed the solution to caculate before, but forgot to remove it.

osmaczko replied:
Would it make sense to make it one field then? Otherwise it is pretty confusing..

qfrank replied:
removed `lastWallTime` @osmaczko

qfrank committed:
chore_: remove field lastWallTime

igor-sirotin commented on timesource/timesource_test.go:
```diff
@@ -305,3 +306,262 @@ func TestGetCurrentTimeOffline(t *testing.T) {
+func TestSystemTimeChangeDetection(t *testing.T) {
+	// Create a controlled time source with fixed time
+	currentTime := time.Now()
+	timeJump := 2 * TimeChangeThreshold
```
```suggestion
	const timeJump = 2 * TimeChangeThreshold
```

qfrank replied:
fixed

igor-sirotin commented on timesource/timesource_test.go:
```diff
@@ -305,3 +306,262 @@ func TestGetCurrentTimeOffline(t *testing.T) {
+
+	// Start the time source
+	ts.Start()
+	// Fix: Handle error from Stop in defer
```
leftover comment? I think you're already handling the error?

qfrank replied:
fixed

igor-sirotin commented on timesource/timesource.go:
```diff
@@ -144,19 +148,53 @@ type NTPTimeSource struct {
+
+	// If significant time change detected, update offset synchronously
+	if timeDiff.Abs() > TimeChangeThreshold {
+		logutils.ZapLogger().Warn("System time change detected",
```
```suggestion
		logutils.ZapLogger().Warn("system time change detected",
```

qfrank replied:
fixed

igor-sirotin commented on timesource/timesource.go:
```diff
@@ -144,19 +148,53 @@ type NTPTimeSource struct {
+			zap.Duration("threshold", TimeChangeThreshold))
+
+		// Ignore error as it's logged in updateOffset
+		_ = s.updateOffset()
```
Is it safe in terms of performance? We query time servers during a simple `now()` function. 
Though.. it should only happen once, and only when the time got changed, so perhaps it's fine.

osmaczko replied:
If the system clock is off, each invocation of `Now()` will result in `updateOffset`, right? If that's the case, that's indeed not good.

igor-sirotin replied:
What do you mean by `system clock is off`? 
If you change the time locally, it will still go with the same pace as _real_ clock. It's just the offset that will change. And that's why I think this check will only go once, as soon as you change the time locally.
But maybe I'm not getting something 🤔

osmaczko replied:
You're right, my brain must’ve already gone to bed. All good!

igor-sirotin reviewed (approved):
Took some time to figure out what's happening, but looks good 👍

qfrank committed:
chore_: address review feedback from Igor
